Initial commit for DMR architecture (#16)
buerokratt/DMR: docs/DMR-design-architecture.doc.md

> # Task

Reviewer: DMR-design-architecture.doc.md -> DMR-technical-architecture.md just to keep everything consistent (https://github.com/buerokratt/Mock-Classifier/tree/main/docs/technical)

## User Story
AS a Developer
I WANT a design for services covering all aspects of managing participants of Bürokratt´s ecosystem and communication between them
SO THAT can I use the design to implement DMR itself and have it communicate with all its related services

This user story is for the expansion of requirements and technical design of the DMR system.

## Acceptance Criteria
- [ ] List all participants involved
- [ ] Cover the flow of managing participants of Bürokratt´s ecosystem
- [ ] Cover the flow where user´s request needs to be classified
- [ ] Cover the flow where classified request gets sent to appropriate bot
- [ ] Cover the flow where a response from a third-party bot ends up at Client´s GUI

# Cross-functional requirements

> - All connections are closed by requester after the request has been made. It means that the requester will not keep up the connection to get the response. A callback URI is provided instead if necessary.

Reviewer: ```suggestion
- Requests between components are short lived, meaning that the connection will usually end with the confirmation of delivery and a callback URI if necessary.
```
Currently sounds like the requester has to drop the connection without actually receiving a response :)

## Participants
<br>

![DMR participants](./images/dmr-participants.editable.png)

| Participant  | Description                                                                                                                                |
|--------------|--------------------------------------------------------------------------------------------------------------------------------------------|
| Client       | End-client´s browser, curl request or similar                                                                                              |
| Chatbot      | Client´s instance of Bürokratt chatbot service that in this example is used directly by Client                                             |
| Chatbots 1-N | Identical to "Chatbot" but is not used by Client in this example                                                                           |
| DMR          | Distributed Message Rooms (N number of them) pass on messages between different chatbot instances and Classifier. Dumb pipe by its nature. |
| CentOps      | Central Operating System controlled by Bürokratt´s core team to track and manage the whole ecosystem of Bürokratt                          |

> | Classifier   | A stand-alone service to detect which participant is the correct one to process given (user) input                                         |

Reviewer: I would prefer not to use the word "ministry" because the end goal is to have a solution that can be used anywhere (including private sector)

> | Classifier   | A stand-alone service to detect which participant is the correct one to process given (user) input                                         |

Author: > I would prefer not to use the word "ministry" because the end goal is to have a solution that can be used anywhere (including private sector)

Same here. But I´ll upload a proper version of it soon, let´s review based on that.

> | Message Que  | Out of scope now, potentially valuable to store information about unsuccessful requests within the whole ecosystem                         |

Reviewer 2: My recommendation is to completely remove Message Queue because it is not something that we are doing right now. The design should just cover what is intended for delivery right now.

# Management
## CentOps-related requests
<br>

![Requests to Centops](./images/requests-to-centops.editable.png)

### Becoming a member of Bürokratt ecosystem
#### Request by Chatbot 

> `https://byk.centops/register-chatbot`

Reviewer: In addition something to keep in mind https://martinfowler.com/articles/richardsonMaturityModel.html
I would really like to have all endpoints being at least level 2, meaning we use resources and proper verbs, i.e. in this case "register-chatbot" is not really a resource but a command, so maybe something like:
POST /participants
{"type": "Chatbot"}

This would make it easier to include new types of participants in the future while keeping the model similar. Also, then the consumers of the API know what to expect, e.g. we can provide endpoints like:
GET /participants/10
GET /participants?type=chatbot
DELETE /participants/10
etc...

```
{
	"client_information": ["name", "short_description", "institution"],
	"initial_verification_token_by_byk": "hash",
	"endpoints": ["public", "for_centops", "for_dmr"],
	"valid": ["from", "to"],
	"participation_in_ecosystem": ["common_cache_use_contribute", "federated_learning_use_contribute", "respond_to_third_party_requests"]
}
```

#### Request by DMR

`https://byk.centops/register-dmr`

```
{
	"client_information": ["name", "short_description"],
	"initial_verification_token_by_byk": "hash",
	"endpoints": ["public", "for_centops", "for_chatbots", "for_classifier"],
	"valid": ["from", "to"]
}
```

> <!--

Reviewer 2: Please remove commented text.

`https://private.dmr/centops-notifications/participation-in-ecosystem`

```
{
	"request_id": "Hash generated and ping-ponged among Chatbot/DMR and CentOps",
	"response": ["in_progress", "accepted", "rejected"]
}
```
-->

### Reporting about planned outages
#### Request by Chatbot / DMR

`https://byk.centops/planned-outage-chatbot`
`https://byk.centops/planned-outagesuspend-dmr`

```
{
	"client_information": ["client_id", "reason_for_outage"],
	"valid": ["from", "to"]
}
```

### Suspending their participation in Bürokratt ecosystem
#### Request by Chatbot / DMR

`https://byk.centops/suspend-chatbot`
`https://byk.centops/suspend-dmr`

```
{
	"client_information": ["client_id", "reason_for_suspending"],
	"valid": ["from", "to"]
}
```

### Reporting about returning to Bürokratt ecosystem
#### Request by Chatbot / DMR

`https://byk.centops/return-chatbot`
`https://byk.centops/return-dmr`

```
{
	"client_information": ["client_id"]
}
```

### Removing themselves from Bürokratt ecosystem
#### Request by Chatbot / DMR

`https://byk.centops/remove-chatbot`
`https://byk.centops/remove-dmr`

```
{
	"client_information": ["client_id", "reason_for_suspending"]
}
```

> ### Request information about software updates

Reviewer: This section raises a few questions. The end goal is to support different types and versions of chatbots participating in the ecosystem right? E.g. byk-chatbot and azure chatbot are both supported as long as they conform to the API requirements?
In that case focusing on software-updates seems very byk base component specific and I would prefer to focus on API version instead.
But if the intention is to actually notify clients about their need to upgrade to newer versions of base components due to security patches etc, then it might sense.
Maybe we can start adding the end goals next to the API as well?

#### Request by Chatbot / DMR

> Different endpoints are used so that restricting access to them could be applied on server instead of application level

`https://byk.centops/list-software-updates/chatbot`
`https://byk.centops/list-software-updates/dmr`

```
{

> 	"list_componentes": ["*", "service.chatbot", "service.dmr", "ruuter", "dmapper", "xtr", "tim"]

Reviewer: "list_componentes" - love the spanish twist :)

}
```

> In case of using `service.` as a prefix, a container containing everything necessary to run a full-blown service is given. Otherwise, only specific components are covered.

#### Response by CentOps

```
{
	"ruuter": {
		"latest": "1.2.3",
		"latest_allowed": "1.2.0"
	}
}
```

### Sending aggregated logs to CentOps

These logs

>     - can be used by every client to monitor their system´s health;

Reviewer 3: System here meaning the common Buerokratt components they're running with?

>     - are used by CentOps to monitor the health of 

Reviewer 3: This sentence trails off a bit...

> Automated continuos process

`centops/accept-logs`

```

> timestamp      component_version request_type         forwarded_id       request_id       request_to             response_code   response_in_ms

Reviewer 3: Should we send along the instance of the chatbot which generates these logs?  It might help from a troubleshooting point of view to understand which instances are running into difficulties.

1646220981996  2.0.1             incoming.response    NBZKMWMMMwm72Mvm   -                -                       200             -
```

## CentOps contacts Chatbot / DMR

> **_TO CLARIFY:_** On this image, "_many optional_ Chatbots 1-N" and "_exactly one_ Chatbot" are making requests to CentOps. Both "Chatbots 1-N" and "Chatbot" should actually be treated the same way but as we treat "Chatbot" separately throughout this document, such difference is used.

<br>

![CentOps notifies about updates](./images/centops-notifies-about-updates.editable.png)

### CentOps notifies Chatbot / DMR about them being accepted as part of Bürokratt ecosystem

`https://private.chatbot/centops-notifications/participation-in-ecosystem`
`https://private.dmr/centops-notifications/participation-in-ecosystem`

```
{
	"request_id": "Hash generated and ping-ponged among Chatbot/DMR and CentOps",
	"response": ["in_progress", "accepted", "rejected"]
}
```

### CentOps notifies about the need to update software

`https://private.chatbot/centops-notifications/request-for-software-update`

> `https://private.dmr/centops-notifications/request-for-software-update`

Reviewer 3: In terms of the Chatbot - I think I understand that a chatbot is a collection of running containers - we'd tell the chat bot maintainer to update a particular container they're running alongside.

In terms of the DMR - which Buerokratt components do we see it using that it hosts itself that would need updating?  Do we see sidecar containers with Buerokratt doing some of the work here?

```
[{
	"participant": ["name", "description", "comments"],
	"labels": "Classifier",
	"uri": "https://private.classifier/input-from-dmr/get-institution-name",
	"status": ["up", "down_unknown", "planned_outage"]
}]
```

### CentOps updates Chatbots and Classifiers with the list of DMRs

`https://private.chatbot/centops-notifications/update-list-of-dmrs`
`https://private.classifier/centops-notifications/update-list-of-dmrs`

> Having Classifier on the picture here although we are not covering it at the moment -- total mock for POC

```
[{
	"participant": ["name", "description", "comments"],
	"labels": "DMR_A",
	"uri": {
		"unclassified": "https://dmr/unclassified",
		"classified": "https://dmr/response-to-request"
	},
	"status": ["up", "down_unknown", "planned_outage"]
}]
```

# Chat flow
## Initial request from the Client
<br>

![DMR participants](./images/client-request.editable.png)

- Client sends a request to any member of Bürokratt´s ecosystem
- Client does not have to know which chatbot is the right one to answer him/her
- Client may ask about different intents within the same chat session knowing (or not) that this particular chatbot is not the one capable of responding

> - Regardles of everything described above, the Client must still get a proper response to their request within this same chat session

Reviewer 2: ```suggestion
- Regardless of everything described above, the Client must still get a proper response to their request within this same chat session
```

### Request from Client to Chatbot

`https://private.chatbot/post-message`

```
{
	"message": "Standard"
}
```

## Chatbot unable to process the request
<br>

![Chatbot to DMR](./images/chatbot-to-dmr.editable.png)

- A local Chatbot that the Client turned to is unable to find an appropriate answer
- In all such cases, an initial request is passed on to DMR which has to know what to do next

### Request from Chatbot to DMR

`https://dmr/unclassified`

```
{
	"callback_uri": "https://private.chat/chat-callback",
	"request_id": "J8EQMgzANpXA7gLnqMERcfcd",
	"payload": "ALLSAsnR4zJ42AyMQWFsq7q9AUvPrgqHHYRrAhxX35eh6QhTPJnbyX7wZW48KgSYaBcuBSEMycwRNbB4MgXpgCHXLuG4R38Z"
}
```

> `payload` can not be read by DMR

## DMR sends the request to Classifier
<br>

![DMR to Classifier](./images/dmr-to-classifier.editable.png)

> - If it´s an initial request (neither classified or responded by another bot), it always gets sent to Classifier

Reviewer 2: ```suggestion
- If it is an initial request (neither classified or responded by another bot), it always gets sent to Classifier
```

### Request from DMR to Classifier

> `https://private.classifier/input-from-dmr/get-institution-name`

Reviewer 2: `get-institution-name` is not a restful endpoint. To be restful, the endpoint should simply be `institutions` and the http verb used will determined the action.

	"institution": "Chatbot_A"
}
```

> In case of multiple institutions, Classifier makes separate requests for all of them

## DMR sends the request to proper participant
<br>

![DMR to third party chatbots](./images/dmr-to-third-party-chatbots.editable.png)

> CentOps continuously provides DMR information about which institution keywords (`Chatbot_A` = `https://private.chatbot_a/message-from-dmr` in this case) correspond to which endpoints

### Request from DMR to proper Chatbot
`https://private.chatbot_a/message-from-dmr`

```
{
	"callback_uri": "https://private.chat/chat-callback",
	"request_id": "J8EQMgzANpXA7gLnqMERcfcd",
	"payload": "ALLSAsnR4zJ42AyMQWFsq7q9AUvPrgqHHYRrAhxX35eh6QhTPJnbyX7wZW48KgSYaBcuBSEMycwRNbB4MgXpgCHXLuG4R38Z",
	"institution": "Chatbot_A"
}
```

> Classifier is able to read the content of `payload`

## Third-party chatbot responds to DMR
<br>

![Third-party chatbot to DMR](./images/third-party-chatbot-to-dmr.editable.png)

- Any third-party chatbot acts as any other chatbot would regardless of if they´re requested by an actual Client, DMR or something else

> - As a request from Chatbot states a mandatori response URI (among other things), the response is sent to any of the DMRs currently active

Reviewer 2: ```suggestion
- As a request from Chatbot states a mandatory response URI (among other things), the response is sent to any of the DMRs currently active
```
